designer/client/inline-conditions.js

```diff
@@ -0,0 +1,200 @@
+import React from 'react'
+import { ConditionsModel, Condition, Field, Value } from './inline-condition-model'
+import { clone } from './helpers'
+
+const conditionalOperators = {
+  'default': {
+    is: {},
+    'is not': {}
+  },
+  'TextField': {
+    matches: {}
+  },
+
+  getConditionals: (fieldType) => {
+    return Object.assign(conditionalOperators[fieldType] || {}, conditionalOperators['default'])
+  }
+}
+
+class InlineConditions extends React.Component {
+  constructor (props) {
+    super(props)
+
+    const { data } = this.props
+
+    this.fields = [].concat.apply([],
+      data.pages.map(page => page.components.filter(component => component.name)
+        .map(component => ({ label: component.title, name: component.name, type: component.type, values: component.options.list ? data.lists.find(it => it.name === component.options.list).items : [] })))
+    ).reduce((obj, item) => {
+      obj[item.name] = item
+      return obj
+    }, {})
+
+    this.state = {
+      conditions: new ConditionsModel()
+    }
+  }
+
+  onClickAddItem = e => {
+    this.setState({
+      condition: {}
+    })
+  }
+
+  onClickAnd = e => {
+    this.setState({
+      condition: {
+        coordinator: 'and'
+      }
+    })
+  }
+
+  onClickOr = e => {
+    this.setState({
+      condition: {
+        coordinator: 'or'
+      }
+    })
+  }
+
+  onClickSave = e => {
+    const { conditions, condition } = this.state
+
+    this.setState({
+      conditions: conditions.add(new Condition(Field.from(condition.field), condition.operator, Value.from(condition.value), condition.coordinator)),
+      condition: undefined
+    })
+  }
+
+  conditionFieldId (id, fieldName) {
+    return `conditions.${id}.${fieldName}`
+  }
+
+  parseConditionId (fieldId) {
+    return fieldId.split('.')[1]
+  }
+
+  onChangeField = e => {
+    const input = e.target
+    const fieldName = input.value
+
+    const { condition } = this.state
+
+    const currentField = condition.field ? condition.field.name : undefined
+    const currentOperator = condition.operator
+
+    this._updateCondition(condition, c => {
+      if (fieldName) {
+        if (currentField && this.fields[currentField].values !== this.fields[fieldName].values) {
+          delete c.value
+        }
+        if (currentOperator && !conditionalOperators.getConditionals(fieldName)[currentOperator]) {
+          delete c.operator
+        }
+        c.field = new Field(fieldName, this.fields[fieldName].label)
+      } else {
+        delete c.field
+        delete c.operator
+        delete c.value
+      }
+    })
+  }
+
+  _updateCondition (condition, updates) {
```

Comment:
I think `updateCondition` should be calling `setState` and changing the state rather than changing class properties. It's hard to tell what this is doing without being able to run it.

Reply:
it calls setState with an updated copy of the class. I thought this would be fine since it's not manipulating the object contained in the state directly and then calls setState with an updated object?

Reply:
@jenbutongit are you happy for me to resolve this one?